src/main/scala/firrtl/annotations/MemoryInitAnnotation.scala

```diff
@@ -33,3 +40,30 @@ case class MemoryArrayInitAnnotation(target: ReferenceTarget, values: Seq[BigInt
   override def initValue:    MemoryInitValue = MemoryArrayInit(values)
   override def isRandomInit: Boolean = false
 }
+
+/** Initialize the `target` memory with inline readmem[hb] statement. */
+case class MemoryFileInlineAnnotation(
+  target:      ReferenceTarget,
+  filename:    String,
+  hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex)
+    extends MemoryInitAnnotation {
+  override def duplicate(n: ReferenceTarget): Annotation = copy(n)
+  override def initValue: MemoryInitValue = {
+
+    if (filename.trim().isEmpty()) {
+      throw new Exception(s"empty filename not allowed in MemoryFileInlineAnnotation")
+    }
+
+    val readmem = hexOrBinary match {
+      case MemoryLoadFileType.Binary => "$readmemb"
+      case MemoryLoadFileType.Hex    => "$readmemh"
+    }
+
+    val inlineLoad = s"""
+                        |initial begin
+                        |  $readmem("$filename", ${target.ref});
+                        |end""".stripMargin
+    MemoryFileInlineInit(inlineLoad)
+  }
```

Comment:
I think it would be better if you could just return a `MemoryFileInlineInit(filename, hexOrBinary)` and move the code that generates the Verilog into the emitter.